src/main/java/org/jboss/logmanager/formatters/Formatters.java

```diff
@@ -477,6 +483,30 @@ public void renderRaw(Formatter formatter, final StringBuilder builder, final Ex
         };
     }
 
+    private static DateTimeFormatter createDateTimeFormatter(String formatString) {
+        if ((formatString == null) || "ISO8601".equals(formatString)) { // Logback uses 'ISO8601'
+            return DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss,SSS");
+        }
+        final DateTimeFormatter cached = CACHED_DATE_TIME_FORMATTERS.get(formatString);
+        if (cached != null) {
+            return cached;
+        }
+        if (CONSTANT_PATTERN.matcher(formatString).matches()) {
+            DateTimeFormatter result;
+            try {
+                final MethodHandle methodHandle = MethodHandles.publicLookup()
+                        .findStaticGetter(DateTimeFormatter.class, formatString, DateTimeFormatter.class);
+                result = (DateTimeFormatter) methodHandle.invoke();
+            } catch (Throwable e) {
+                StandardOutputStreams.printError("Failed to determine date constant %s. Defaulting to %s.", formatString, "yyyy-MM-dd HH:mm:ss,SSS");
+                result = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss,SSS");
+            }
+            CACHED_DATE_TIME_FORMATTERS.put(formatString, result);
```

Comment:
I recall seeing somewhere that ThreadLocal was used instead of a `ConcurrentHashMap`. Does that still exist? Is it a concern here?

Reply:
IIRC a ThreadLocal was used because we used the `SimpleDateFormatter` which is not thread safe. The `DateTimeFormatter` should be thread safe, so  think it should work. TBH we might not even need to cache it. It's likely not invoked that often at all.

Reply:
That's why I am asking, I don't know how often this is called :)